### PR TITLE
Palindrome benchmark

### PR DESCRIPTION
**Pull request by ggreif**
This PR ports the `palindrome` algorithms from

https://www.brics.dk/RS/05/3/BRICS-RS-05-3.pdf

The idea is that this microbenchmark will test the _call_, _closure forming_ and _pattern-matching_ performance.

### CODE REVIEW

File: test/bench/palindrome.mo [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    func direct(xs : List<Char>) : Bool {
+        func walk (xs1 : List<Char>, xs2 : List<Char>) : List<Char> =
+            switch (xs1, xs2) {
+                case (_, null) xs1; // even length
+                case (?(_, xs1), ?(_, null)) xs1; // odd length
+                case (?(x, xs1), ?(_, ?(_, xs2))) {
+                    let ?(y, ys) = walk (xs1, xs2);
+                    if (x == y) {
+                        ys
+                    } else { throw error("Nope") }
```
Comment (ggreif) [inline test/bench/palindrome.mo:30]: @crusso I get error messages from here, no idea why...

Comment (crusso) [reply test/bench/palindrome.mo:30]: You can't try or throw from a non async context.

File: test/bench/palindrome.mo [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    func cps(xs : List<Char>) : Bool {
+        func walk(xs1 : List<Char>, xs2 : List<Char>, k : List<Char> -> Bool) : Bool =
+            switch (xs1, xs2) {
+                 case (_, null) { k xs1 };
+                 case (?(_, xs1), ?(_, null)) { k xs1 };
+                 case (?(x, xs1), ?(_, ?(_, xs2))) {
+                     walk (xs1, xs2, func (?(y, ys)) = x == y and k ys)
+                 }
+             };
+        walk (xs, xs, func null = true)
```
Comment (crusso) [inline test/bench/palindrome.mo:18]: `func null = true` will trap on non-empty list - don't you want to return false instead?
`func l { l == null;}` might work better.

File: test/bench/palindrome.mo (hunk 1 above)
Comment (ggreif) [reply test/bench/palindrome.mo]: Rewritten to use `do ?` instead.

File: test/bench/palindrome.mo (hunk 2 above)
Comment (ggreif) [reply test/bench/palindrome.mo:18]: No, this works and is as close to the original as possible.

File: test/bench/palindrome.mo [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    };
+
+    func direct(xs : List<Char>) : Bool {
+        func walk (xs1 : List<Char>, xs2 : List<Char>) : List<Char> =
```
[the remark this answers is not in the document]
Comment (ggreif) [reply test/bench/palindrome.mo]: It does now.

**Review by crusso — COMMENTED** [review]
Indentation is a bit funny, but here's a rubber stamp.

**Review by crusso — APPROVED** [review]
(rubber stamping)

**Comment by ggreif** [comment]
> Indentation is a bit funny, but here's a rubber stamp.

Emacs cannot indent `do ?` properly.

**Pull request merged**